wallet/wallet-rpc-lib/src/rpc/mod.rs

```diff
@@ -1274,19 +1277,104 @@ impl<N: NodeInterface + Clone + Send + Sync + 'static> WalletRpc<N> {
+        htlc: RpcHashedTimelockContract,
+        config: ControllerConfig,
+    ) -> WRpcResult<SignedTransaction, N> {
+        let secret_hash = HtlcSecretHash::decode_all(&mut htlc.secret_hash.as_bytes())
+            .map_err(|_| RpcError::InvalidHtlcSecretHash)?;
+
+        let spend_key = htlc
+            .spend_address
+            .decode_object(&self.chain_config)
+            .map_err(|_| RpcError::InvalidAddress)?;
+
+        let refund_key = htlc
+            .refund_address
+            .decode_object(&self.chain_config)
+            .map_err(|_| RpcError::InvalidAddress)?;
+
+        let htlc = HashedTimelockContract {
+            secret_hash,
+            spend_key,
+            refund_timelock: htlc.refund_timelock,
+            refund_key,
+        };
+
+        let token_id = match token_id {
+            Some(id) => {
+                Some(id.decode_object(&self.chain_config).map_err(|_| RpcError::InvalidTokenId)?)
+            }
+            None => None,
+        };
+        let coin_decimals = self.chain_config.coin_decimals();
+
+        self.wallet
+            .call_async(move |controller| {
+                Box::pin(async move {
+                    let value = match token_id {
+                        Some(token_id) => {
+                            let token_info = controller.get_token_info(token_id).await?;
+                            let amount = amount
+                                .to_amount(token_info.token_number_of_decimals())
+                                .ok_or(RpcError::InvalidCoinAmount)?;
+                            OutputValue::TokenV1(token_id, amount)
+                        }
+                        None => {
+                            let amount = amount
+                                .to_amount(coin_decimals)
+                                .ok_or(RpcError::InvalidCoinAmount)?;
+                            OutputValue::Coin(amount)
+                        }
+                    };
+
+                    controller
+                        .synced_controller(account_index, config)
+                        .await?
+                        .create_htlc_tx(value, htlc)
+                        .await
+                        .map_err(RpcError::Controller)
+                })
+            })
+            .await?
+    }
+
     pub async fn compose_transaction(
         &self,
         inputs: Vec<RpcUtxoOutpoint>,
         outputs: Vec<TxOutput>,
+        htlc_secrets: Option<Vec<Option<RpcHexString>>>,
         only_transaction: bool,
     ) -> WRpcResult<(TransactionToSign, Balances), N> {
         ensure!(!inputs.is_empty(), RpcError::ComposeTransactionEmptyInputs);
         let inputs = inputs.into_iter().map(|o| o.into_outpoint()).collect();
+
+        let htlc_secrets = htlc_secrets
+            .map(|s| {
+                s.into_iter()
+                    .map(|s| {
+                        s.map(|s| -> Result<HtlcSecret, RpcError<N>> {
```

Comment:
Choosing different names instead of s would make the code a little more readable. E.g. `secrets_opt`, `secret_opt`, `secret`.